Update Maddy to not convert internal single underscores.

luciansmith opened the pull request:
Still has problems with internal double-underscores, but at least those are less common.

I've been going through all the 'notes' in Biomodels models, checking to see if they roundtrip through the new markdown parsers.  Many don't!  I've been submitting issue requests and fixes to maddy and to html2md.  This is a pretty large fix, as there are many URLs with internal underscores in them that maddy was converting to `<em>`.

However, there are other problems; if you have to release now, we do indeed do what we claim (use those two libraries to convert markdown); it's just that those libraries have issues.

luciansmith committed:
Update Maddy to not convert internal single underscores.

Still has problems with internal double-underscores, but at least those are less common.

fbergmann commented:
@luciansmith a release right now would at least help with the issues of broken python3.13 bindings, FBC issues, and the other fixes over the past year or so. We can certainly have another release whenever the MD support has increased further. But i'm happy to pull the 5.20.5 draft release for now.

luciansmith commented:
If a release means that other things get fixed, that's fine!  If this fix could get squeezed in, that'd be great, but it's not crucial.  Since there will be other fixes in the pipeline, if this fix gets bundled with those, that's fine.

luciansmith committed:
Sanitize HTML input for html2md.

Since html2md messes up links if the HTML is somewhat non-standard, sanitizing the input before converting lets the library perform as it should (on the input I've tested it on, at least).

luciansmith committed:
Encode all &, <, and > in markdown.

This is non-standard markdown!  But it has the advantage of creating valid HTML more often, so I think on balance it's a win?

This means that this translator will *not* translate valid HTML in markdown to actual HTML.  I figured this was a more uncommon occurrence than (say) all the biomodels notes that have "<=>" as a reaction symbol, or that have '&' in a URL.

We can drop this (I think) if https://github.com/tim-gromeyer/html2md/issues/158 is addressed.  But Maddy might need updating as well.

Anyway: hack for now, but happy to revert if people think the cost is too severe.

fbergmann reviewed (approved):
@luciansmith this looks good to me, shall i merge it in and release?

luciansmith commented:
Yeah, let's do it.  Thank you!
